src/main/java/com/linecorp/armeria/client/SessionProtocolNegotiationCache.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static java.util.Objects.requireNonNull;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.EnumSet;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.util.LruMap;
+
+/**
+ * Keeps the recent {@link SessionProtocol} negotiation failures.
+ */
+public final class SessionProtocolNegotiationCache {
+
+    private static final Logger logger = LoggerFactory.getLogger(SessionProtocolNegotiationCache.class);
+
+    private static final Map<String, CacheEntry> cache = new LruMap<String, CacheEntry>(65536) {
+        private static final long serialVersionUID = -2506868886873712772L;
+
+        @Override
+        protected boolean removeEldestEntry(Entry<String, CacheEntry> eldest) {
+            final boolean remove = super.removeEldestEntry(eldest);
+            if (remove) {
+                logger.debug("Evicted: '{}' does not support ", eldest.getKey(), eldest.getValue());
+            }
+
+            return remove;
+        }
+    };
+
+    public static boolean isUnsupported(SocketAddress remoteAddress, SessionProtocol protocol) {
+        final String key = key(remoteAddress);
+        final CacheEntry e;
+        synchronized (cache) {
```

Comment:
After thinking a little bit more, I'm leaning even more toward leaving this as is. This is called only when making a new connection attempt and our connections live long.